openhands/llm/retry_mixin.py

```diff
@@ -0,0 +1,85 @@
+from tenacity import (
+    retry,
+    retry_if_exception_type,
+    stop_after_attempt,
+    wait_exponential,
+)
+
+from openhands.core.exceptions import OperationCancelled
+from openhands.core.logger import openhands_logger as logger
+from openhands.runtime.utils.shutdown_listener import should_exit
+
+
+class RetryMixin:
+    """Mixin class for retry logic."""
+
+    def retry_decorator(self, **kwargs):
+        """
+        Create a retry decorator with customizable parameters.
+
+        Args:
+            **kwargs: Keyword arguments to override default retry behavior.
+                      Keys: num_retries, retry_exceptions, retry_min_wait, retry_max_wait, retry_multiplier
+
+        Returns:
+            A retry decorator with the specified or default parameters.
+        """
+        num_retries = kwargs.get('num_retries')
+        retry_exceptions = kwargs.get('retry_exceptions')
+        retry_min_wait = kwargs.get('retry_min_wait')
+        retry_max_wait = kwargs.get('retry_max_wait')
+        retry_multiplier = kwargs.get('retry_multiplier')
+
+        def custom_completion_wait(self, retry_state):
+            """Custom wait for completion."""
+            exception = retry_state.outcome.exception() if retry_state.outcome else None
+            if exception is None:
+                return 0
+
+            # retry_min_wait = config.retry_min_wait
+            # retry_max_wait = config.retry_max_wait
+
+            # rate limit errors
+            exception_type = type(exception).__name__
+            logger.error(rf'\exception_type: {exception_type}\n')
+
+            if exception_type == 'RateLimitError':
+                retry_min_wait = 60
+                retry_max_wait = 240
+            elif exception_type == 'BadRequestError' and exception.response:
+                # this should give us the buried, actual error message
+                # from the LLM model.
+                logger.error(f'\n\nBadRequestError: {exception.response}\n\n')
```

Comment:
We need to retry transitory errors, but BadRequestError is typically a permanent error. If we retry, several times, permanent errors, then what happens is: the user is waiting, the _rate limits_ add up numbers of retries, we struggle to debug / log. And it's for no result: if it's a 400, like those related to format of the request, it will not succeed. It still fails. So the retry spent user's time and did nothing.

On the other hand, the need for logging here is completely understandable, of course we need to look into the original exception. Once we do [this revert](https://github.com/All-Hands-AI/OpenHands/pull/4054) though, we have them: 400 will get [here](https://github.com/All-Hands-AI/OpenHands/blob/9241ae2148dc922cfdf87c91a7ba99533cde4581/openhands/controller/agent_controller.py#L160).

Reply:
That was a struggle with litellm when I tried looking into I think Gemini erroring on the syntax with the image_url's, iirc.
If this does some better logging, maybe keep it, otherwise remove it.